lib/Push.php

```diff
@@ -0,0 +1,179 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Notifications;
+
+
+use OC\Authentication\Exceptions\InvalidTokenException;
+use OC\Authentication\Token\IProvider;
+use OC\Security\IdentityProof\Key;
+use OC\Security\IdentityProof\Manager;
+use OCP\AppFramework\Http;
+use OCP\Http\Client\IClientService;
+use OCP\IConfig;
+use OCP\IDBConnection;
+use OCP\ILogger;
+use OCP\IUser;
+use OCP\IUserManager;
+use OCP\Notification\IManager;
+use OCP\Notification\INotification;
+
+class Push {
+	/** @var IDBConnection */
+	protected $connection;
+	/** @var IManager */
+	protected $manager;
+	/** @var IConfig */
+	protected $config;
+	/** @var IProvider */
+	protected $tokenProvider;
+	/** @var Manager */
+	private $keyManager;
+	/** @var IUserManager */
+	private $userManager;
+	/** @var IClientService */
+	protected $clientService;
+	/** @var ILogger */
+	protected $log;
+
+	public function __construct(IDBConnection $connection, IManager $manager, IConfig $config, IProvider $tokenProvider, Manager $keyManager, IUserManager $userManager, IClientService $clientService, ILogger $log) {
+		$this->connection = $connection;
+		$this->manager = $manager;
+		$this->config = $config;
+		$this->tokenProvider = $tokenProvider;
+		$this->keyManager = $keyManager;
+		$this->userManager = $userManager;
+		$this->clientService = $clientService;
+		$this->log = $log;
+	}
+
+	/**
+	 * @param INotification $notification
+	 */
+	public function pushToDevice(INotification $notification) {
+		$devices = $this->getDevicesForUser($notification->getUser());
+		$user = $this->userManager->get($notification->getUser());
+
+		if (empty($devices) || !($user instanceof IUser)) {
+			return;
+		}
+
+		$language = $this->config->getUserValue($notification->getUser(), 'core', 'lang', 'en');
+		try {
+			$notification = $this->manager->prepare($notification, $language);
+		} catch (\InvalidArgumentException $e) {
+			return;
+		}
+
+		$userKey = $this->keyManager->getKey($user);
+
+		$pushNotifications = [];
+		foreach ($devices as $device) {
+			try {
+				$pushNotifications[] = $this->encryptAndSign($userKey, $device, $notification);
+			} catch (InvalidTokenException $e) {
+				// Token does not exist anymore, should drop the push device entry
+				// FIXME delete push token
+			} catch (\InvalidArgumentException $e) {
+				// Token does not exist anymore, should drop the push device entry
+				// FIXME delete push token
+			}
+		}
+
+		$client = $this->clientService->newClient();
+		try {
+			$pushServer = rtrim($this->config->getAppValue('notifications', 'push_server', 'https://push-notifications.nextcloud.com'), '/');
```

Comment:
@LukasReschke please also confirm the URL where we are going to place our push proxy